benchmark/index.ts

```diff
@@ -0,0 +1,200 @@
+	fn: async (deferred: {resolve(): void}) => {
+		const stream = new Request(url, gotOptions);
+		stream.resume().once('end', () => {
+			deferred.resolve();
+		});
+	}
+}).add('got - stream core - normalized options', {
+	defer: true,
+	fn: async (deferred: {resolve(): void}) => {
+		const stream = new Request(undefined as any, normalizedGotOptions);
+		stream.resume().once('end', () => {
+			deferred.resolve();
+		});
+	}
+}).add('request - callback', {
+	defer: true,
+	fn: (deferred: {resolve(): void}) => {
+		request(urlString, requestOptions, (error: Error) => {
+			if (error) {
+				throw error;
+			}
+
+			deferred.resolve();
+		});
+	}
+}).add('request - stream', {
+	defer: true,
+	fn: (deferred: {resolve(): void}) => {
+		const stream = request(urlString, requestOptions);
+		stream.resume();
+		stream.once('end', () => {
+			deferred.resolve();
+		});
+	}
+}).add('node-fetch - promise', {
+	defer: true,
+	fn: async (deferred: {resolve(): void}) => {
+		const response = await fetch(url, fetchOptions);
+		await response.text();
+
+		deferred.resolve();
+	}
+}).add('node-fetch - stream', {
+	defer: true,
+	fn: async (deferred: {resolve(): void}) => {
+		const {body} = await fetch(url, fetchOptions);
+
+		body.resume();
+		body.once('end', () => {
+			deferred.resolve();
+		});
+	}
+}).add('axios - promise', {
+	defer: true,
+	fn: async (deferred: {resolve(): void}) => {
+		await axios.request(axiosOptions);
+		deferred.resolve();
+	}
+}).add('axios - stream', {
+	defer: true,
+	fn: async (deferred: {resolve(): void}) => {
+		const {data} = await axios.request(axiosStreamOptions);
+		data.resume();
+		data.once('end', () => {
+			deferred.resolve();
+		});
+	}
+}).add('https - stream', {
+	defer: true,
+	fn: (deferred: {resolve(): void}) => {
+		https.request(urlString, httpsOptions, response => {
+			response.resume();
+			response.once('end', () => {
+				deferred.resolve();
+			});
+		}).end();
+	}
+}).on('cycle', (event: Benchmark.Event) => {
+	console.log(String(event.target));
+}).on('complete', function (this: any) {
+	console.log(`Fastest is ${this.filter('fastest').map('name') as string}`);
+
+	internalBenchmark();
+}).run();
+
+const internalBenchmark = (): void => {
+	console.log();
+
+	const internalSuite = new Benchmark.Suite();
+	internalSuite.add('got - normalize options', {
+		fn: () => {
+			normalizeArguments(url, gotOptions);
+		}
+	}).on('cycle', (event: Benchmark.Event) => {
+		console.log(String(event.target));
+	});
+
+	internalSuite.run();
+};
+
+// Results (i7-7700k, CPU governor: performance):
+// got - promise                          x 3,092 ops/sec ±5.25% (73 runs sampled)
+// got - stream                           x 4,313 ops/sec ±5.61% (72 runs sampled)
+// got - promise core                     x 6,756 ops/sec ±5.32% (80 runs sampled)
+// got - stream core                      x 6,863 ops/sec ±4.68% (76 runs sampled)
+// got - stream core - normalized options x 7,960 ops/sec ±3.83% (81 runs sampled)
```

Comment:
Normalizing cuts out ~1k req/s but the time is needed to duplicate the options, so the original ones won't be altered if you use hooks to modify the request.